### ISSUE TITLE
Menu: link to fragment identifiers (in page navigation)

### ISSUE THREAD

**Issue by olafgrabienski** [backdrop/backdrop-issues#4063]
To define menu links to [fragment identifiers](https://www.w3.org/TR/html401/intro/intro.html#fragment-uri) (locations within the same page) for in page navigation on a one pager site, it's not possible to only declare the ID in the `path` input (validation error). The trick is to declare the page before the ID.

- Doesn't work: `#about`
- Works well: `<front>#about`, `node/123#about`

---
**Original issue description**

For a one pager site, I'd like to define menu links to [fragment identifiers](https://www.w3.org/TR/html401/intro/intro.html#fragment-uri) (locations within the same page) for in page navigation.

Example:
```
<ul class="menu-dropdown" ...>
  <li><a href="#about">About</a></li>
  <li><a href="#features">Features</a></li>
  <li><a href="#contact">Contact</a></li>
</ul>
```

That's currently not possible. When I put `#about` in the *Path* input of a menu link, I get a message about an invalid path and can't save the menu item.

**Proposed solution**
Change validation to allow links to fragment identifiers.

**Alternatives that have been considered**
(1) Use of complete URLs including fragment identifiers, e.g. `https://example.com#about`.
  Drawback: Doesn't work well if you have several versions of a site (dev and production).

(2) Build a custom menu structure as HTML in a block.
  Drawbacks: Editors can't use the menu system, no Smartmenu toggle functionality etc.

**Comment by olafgrabienski** [comment]
For reference, there is a similar request for link fields: #1308

**Comment by indigoxela** [comment]
There's a trick: use something like `node#main-content` or `node/1#main-content`.
(#main-content just as an example here).

Advantage: if you later add more pages, it all still works. And "node/XX" always gets the latest alias.

**Comment by olafgrabienski** [comment]
@indigoxela Thanks for this hint, very helpful!

As all my content is on the home page, I don't want to see something like `node` or a path alias in the URL, so I tried to use `<front>#about`, and that works as well, great! The result in the browser address bar is `https://example.com/#about`. (I'm okay with the slash.)

**Comment by olafgrabienski** [comment]
Hm, not sure: keep the issue as a feature request, or change it to `type - documentation`?

**Comment by BWPanda** [comment]
I think 'documentation' (maybe even just this issue is documentation enough) as you'll always need to link a menu item to a specific page (even if you want to link to a specific section of that page). You found the correct solution to first specifying the page (`<front>`) then adding the identifier after that.

I was going to suggest making this a feature request to add this ability (more easily), but I don't think it's a 80% use-case for core. So either a contrib module can do that, or just explain to people how to do it was you've done.

**Comment by olafgrabienski** [comment]
@BWPanda Thanks for your feedback! I've changed the label to `type - documentation` and changed the issue description. Do we leave such documentation issues open, or is it better to close it?

Thanks again @indigoxela for the excellent information!

**Comment by BWPanda** [comment]
I think close. It's resolved, you've found your answer, and others can still find closed issues if they have the same problem.

**Comment by klonos** [comment]
Reopening this one...

So you can for example create menus with these paths:

- `some/path`
- `some/path#scroll-to-footer`

...but not if you specify only `#scroll-to-footer` as path (which is meant to work in **any page** this menu is shown).

**Comment by klonos** [comment]
> The trick is to declare the page before the ID.

Nope, this does not address all use cases. Sometimes no specific page/path needs to be specified. Good/popular examples for this are scroll-to-top/bottom links.

See this article here: https://www.drupal.org/node/106821

It is suggested that this can only be achieved by using contrib modules (such as https://drupal.org/project/menu_token) and/or by writing code. We should allow this from the admin UI.

**Comment by klonos** [comment]
...in Drupal-land, https://www.drupal.org/project/drupal/issues/123103 (NW at the time of writing) seems to be currently blocked by this issue https://www.drupal.org/project/drupal/issues/759808 (NR at the time of writing)
